spock-specs/src/test/groovy/org/spockframework/docs/extension/ParameterInjectionSpec.groovy

```diff
@@ -0,0 +1,64 @@
+package org.spockframework.docs.extension
+
+import org.spockframework.runtime.SpockExecutionException
+import org.spockframework.runtime.extension.ExtensionAnnotation
+import org.spockframework.runtime.extension.IAnnotationDrivenExtension
+import org.spockframework.runtime.extension.ParameterResolver
+import org.spockframework.runtime.model.ParameterInfo
+import spock.lang.Specification
+
+import java.lang.annotation.ElementType
+import java.lang.annotation.Retention
+import java.lang.annotation.RetentionPolicy
+import java.lang.annotation.Target
+
+class ParameterInjectionSpec extends Specification {
+
+  def setupSpec(@ParameterIndex int param1) {
+    assert param1 == 0
+  }
+
+  // tag::example[]
+  def setup(@ParameterIndex int param1) {
+    assert param1 == 0
+  }
+
+  def "test"(@ParameterIndex int param1, @ParameterIndex Integer param2, @ParameterIndex int param3) {
+    expect:
+    param1 == 0
+    param2 == 1
+    param3 == 2
+  }
+  // end::example[]
+
+  def cleanup(@ParameterIndex int param1) {
+    assert param1 == 0
+  }
+
+  def cleanupSpec(@ParameterIndex int param1) {
+    assert param1 == 0
+  }
+}
+
+
+// tag::extension[]
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.PARAMETER)
+@ExtensionAnnotation(ParameterIndexExtension)
+@interface ParameterIndex {}
+
+class ParameterIndexExtension implements IAnnotationDrivenExtension<ParameterIndex> {
+  @Override
+  void visitParameterAnnotation(ParameterIndex annotation, ParameterInfo parameter) {
+    Class<?> type = parameter.reflection.type
+    if (type !in [int, Integer]) {
```

Comment:
```suggestion
    if (!type in [int, Integer]) {
```

Reply:
Why do you think this is better? I think it reads quite nicely `type not in`, with the prefix I would assume on first glance that type somehow gets negated and turned into a boolean.

Reply:
Because of the [compilation errors](https://github.com/spockframework/spock/actions/runs/4387302731/jobs/7682438251#step:4:134) with Groovy 2.5 generated by the former ;-).

A quick check with the Groovy console suggests that my syntax is supported also by Groovy 2.5.

Reply:
They are not the same though, you are missing the parenthesis see [here](https://gwc-experiment.appspot.com/?g=groovy_3_0&codez=eJxLSU1TKKksSFWwVQguKcrMS-fiykxT0AALKWbmKURn5pXoKHjmlaSmpxbFaipUcykAQQFQZUlOnoJSXn6JAlCVoRJXLUSjIlgn0RqNEBo1cOjEpdUYrFUBCQAADZc7Pw)

Reply:
Nice, I didn't think that `!` has higher precedence than in. Good you determined that.